indexing-service/src/main/java/org/apache/druid/indexing/common/actions/CommitRealtimeSegmentsAndMetadataAction.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.common.actions;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.base.Preconditions;
+import org.apache.druid.indexing.common.task.IndexTaskUtils;
+import org.apache.druid.indexing.common.task.Task;
+import org.apache.druid.indexing.overlord.CriticalAction;
+import org.apache.druid.indexing.overlord.DataSourceMetadata;
+import org.apache.druid.indexing.overlord.SegmentPublishResult;
+import org.apache.druid.segment.SegmentUtils;
+import org.apache.druid.timeline.DataSegment;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Task action to commit realtime segments and metadata when using APPEND task locks.
+ * <p>
+ * This action performs the following operations in a single transaction:
+ * <ul>
+ * <li>Commit the append segments</li>
+ * <li>Upgrade the append segments to all visible REPLACE versions</li>
+ * <li>Commit start and end {@link DataSourceMetadata}.</li>
+ * </ul>
+ * This action differs from {@link SegmentTransactionalInsertAction} as it is used
+ * only with APPEND locks and also upgrades segments as needed.
+ */
+public class CommitRealtimeSegmentsAndMetadataAction implements TaskAction<SegmentPublishResult>
```

Comment:
Could we not re-use SegmentTransactionAppendAction with the metadata being null for batch and the required values for streaming ingestion (similar to the original insert action)?

Is there anything besides the metadata commit that this action does that the transactional append action doesn't?

Reply:
Yeah, I was thinking the same. Let me see what we can do.